### PR TITLE
Error checking in select API

### PR DESCRIPTION
**Pull request by abotteram**
## Description

Updated select function to print an error when the called selector does not exist.

The reason I created this PR is because I want to make sure that plugins using the Gutenberg API will not interupt the javascript execution as much as possible. I also believe all other public APIs should have some error checking.

## How Has This Been Tested?

Created unit tests and tested the select API in the browser.
Also improved unit tests to reload the `data/index.js` module for each test, so all module globals are reset, ensuring a clean test environement.

## Types of changes

Improvement.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by aduth** [comment]
Noting that I'd like to explore assigning selector functions into an object produced by the `select` function, rather than specifying them as a second string argument of the function, which conflicts with the changes being proposed here.

**Comment by Xyfi** [comment]
I wan't aware that the current api interface was still open for changes. I would like to point out, although it isn't pretty, the changes I made to the tests will allow for more consistent testing.

File: data/index.js [hunk 1]
```diff
@@ -119,5 +119,12 @@ export const query = ( mapSelectorsToProps ) => ( WrappedComponent ) => {
  * @return {*} The selector's returned value.
  */
 export const select = ( reducerKey, selectorName, ...args ) => {
+	const selector = get( selectors, [ reducerKey, selectorName ] );
+	if ( typeof selector !== 'function' ) {
+		// eslint-disable-next-line no-console
+		console.error( `Invalid selector called, with name "${ selectorName }" for reducer "${ reducerKey }".` +
+			' Make sure the reducerName and selectorName are correct!' );
+		return undefined;
+	}
```
Comment (youknowriad) [inline data/index.js]: Should we add the same check line 103 or maybe reuse the same function?

File: data/test/index.js [hunk 2]
```diff
@@ -38,6 +55,33 @@ describe( 'select', () => {
...
 		expect( selector2 ).toBeCalledWith( store.getState() );
 	} );
+
+	it( 'prints an error when a selector is called on a non-existing reducer', () => {
+		/* eslint-disable no-console */
+		const originalConsoleError = console.error;
+		console.error = jest.fn( () => {} );
+
+		expect( select( 'reducer1', 'selector1' ) ).toBeUndefined();
+		expect( console.error ).toHaveBeenCalled();
```
Comment (youknowriad) [inline data/test/index.js]: Should we use https://github.com/WordPress/packages/tree/master/packages/jest-console instead?

File: data/index.js (hunk 1 above)
Comment (Xyfi) [reply data/index.js]: Fixed.

File: data/index.js [hunk 3]
```diff
@@ -100,10 +100,6 @@ export const query = ( mapSelectorsToProps ) => ( WrappedComponent ) => {
 	};
 
 	return connectWithStore( ( state, ownProps ) => {
-		const select = ( key, selectorName, ...args ) => {
-			return selectors[ key ][ selectorName ]( state[ key ], ...args );
```
Comment (Xyfi) [inline data/index.js]: I just noticed that #4823 broke this line, because `getState()` is not called on `state[ key ]`.

**Comment by gziolo** [comment]
Is it still relevant? There were a few major changes in the `data` codebase.

**Comment by gziolo** [comment]
By the way, we discussed that it might be a good idea to add some mechanism to prevent components to explode when `withSelect` or `withDispatch` try to access a method that isn't there. I'm not sure how challenging it is with the current code for `select` and `dispatch` because they simply return an object :)

**Comment by aduth** [comment]
We could always add a `try` / `catch` around the call to `mapStateToProps` (should really be called `mapSelectToProps`) and `mapDispatchToProps`:

https://github.com/WordPress/gutenberg/blob/85cd6f5988002ae080399c81b34df5a07641ec3e/data/index.js#L278

https://github.com/WordPress/gutenberg/blob/85cd6f5988002ae080399c81b34df5a07641ec3e/data/index.js#L333

**Comment by gziolo** [comment]
> We could always add a try / catch around the call to mapStateToProps (should really be called mapSelectToProps) and mapDispatchToProps:

Yes, totally on board with the idea 💯 Yes, let's rename as suggested 👍

**Comment by gziolo** [comment]
By the way it, would be a good idea to have a helper similar to `deprecated` which would help to standardize the hint for the plugin developer, including the name of the plugin which caused issues.

**Pull request closed without merging**